code/contracts/ONEWallet.sol

```diff
@@ -117,35 +118,66 @@ contract ONEWallet is TokenTracker {
         return nonces[index];
     }
 
-    function getCommits() external view returns (bytes32[] memory, bytes32[] memory, uint32[] memory, bool[] memory)
+    function getCommits() external pure returns (bytes32[] memory, bytes32[] memory, uint32[] memory, bool[] memory){
+        revert("Deprecated");
+    }
+
+    function getAllCommits() external view returns (bytes32[] memory, bytes32[] memory, bytes32[] memory, uint32[] memory, bool[] memory)
     {
-        bytes32[] memory paramHashes = new bytes32[](commits.length);
-        bytes32[] memory hashes = new bytes32[](commits.length);
-        uint32[] memory timestamps = new uint32[](commits.length);
-        bool[] memory completed = new bool[](commits.length);
+        uint32 numCommits = 0;
         for (uint32 i = 0; i < commits.length; i++) {
-            Commit storage c = commitLocker[commits[i]];
+            Commit[] storage cc = commitLocker[commits[i]];
+            numCommits += uint32(cc.length);
+        }
+        bytes32[] memory hashes = new bytes32[](numCommits);
+        bytes32[] memory paramHashes = new bytes32[](numCommits);
+        bytes32[] memory verificationHashes = new bytes32[](numCommits);
+        uint32[] memory timestamps = new uint32[](numCommits);
+        bool[] memory completed = new bool[](numCommits);
+        uint32 index = 0;
+        for (uint32 i = 0; i < commits.length; i++) {
+            Commit[] storage cc = commitLocker[commits[i]];
+            for (uint32 j = 0; j < cc.length; j++) {
+                Commit storage c = cc[j];
+                hashes[index] = c.hash;
+                paramHashes[index] = c.paramsHash;
```

Comment:
Commit structure contains paramHash, which should hide the details of the operation and thus ensure privacy. However, this is not met in all the cases, since paramHash might contain a repeated transfer to the same address with the same amount or the operation with a single parameter (recovery), which can be easily determined by the attacker. It can be resolved by adding some random 32B salt into paramHash, generated by the client.

Anyway, I still think that there is no reason for hiding the operation details in commit:
1) confirmation by OTPs is supposed to be used only for small amounts, so the attacker will not pay for overturning the blockchain to make some double-spending.
2) if you compare it with the native transfer of ONE, signed by ECDSA, the details of the transfer are leaked as well, so the MITM attacker can eclipse the victim in the same manner as in the case of OTPs.

Reply:
I agree in parts. A salt seems reasonable. For the later part regarding reasons to hide operations at commit time, note a distinction between the situation here and that of ECDSA transfer in a single transaction:

1. There is one block confirmation difference (ECDSA transaction would be executed in the same block, here the reveal only happens in the next block).
2. Gas is already paid for the commit (which will be deducted from the wallet, as soon as Account Abstraction is implemented) prior to the occurrence of the fund-transfer transaction. 

These two factors made it very easy for an opponent (in an application which the user is competing with others) to act in response to the user's intent (consider auctions, gamblings, games, etc.). Previously (or in the ECDSA case) the opponent has to operate a very active mining (validation) node, actively scan and parse all transactions in the current block, and pay a much higher gas if the attacker wants to utilize the revealed information to do anything (such as changing price for an auction, place a slightly higher bid, changing odds in a game the attacker hosts, or make advanced movements in a game, etc.). Now, with the parameters revealed in commit, the attacker can just lay back and examine a block that is already confirmed and index where the commit resides.